### ISSUE TITLE
new rule: no-hungarian

### ISSUE THREAD

**Issue by tommywilkinson** [microsoft/tslint-microsoft-contrib#185]
Let's start with the simplest possible form of the rule: don't allow the raw type name to appear in a variable name. It's redundant information and the world has moved on from hungarian naming conventions. We'd want to only compare the class level type name, rather than the full module name since I _hope_ even the biggest hungarian proponents wouldn't go that far. 

``` typescript
// Disallowed
var fooBar: Bar = null

// Allowed
var foo: Bar = null
```

For now I think it makes sense to just look at variable names, but it could possibly be expanded to include functions too.

**Comment by HamletDRC** [comment]
@tommywilkinson Thanks for the feature request. I appreciate the effort you took to write and submit this! 

Did you know that Hungarian notation was invented by a Microsoft programmer on the Word team (his name was Charles Simonyi). It is wonderful that a feature request to ban Hungarian notation is now being submitted to an Open Source Microsoft repository decades later! Time marches forward :)

A few questions: 
1. If you have an object called "AppController" what do you call an instance of it? I think many people would call this "appController". 
2. If you have an object that controls a master-detail view (or parent-child view) with two fields (a MasterController instance and a DetailsController instance) then what do you call those two fields? I think many people would call these "masterController" and "detailsController".

(Please Note... I am not implying that we should base our best practices on a popularity contest of what "most people" would do! I would just like to hear your answers from your team's persepctive). 

3. Would you turn on this rule for your project? How big is the project? Do you think it is too strict? And how many suppressions/edge cases would you probably have? 

4. Have you read Joel's article on Hungarian notation? Search for "Hungarian" to get to the good part. Apps Hungarian can be useful but Systems Hungarian is pointless. However, your description of Hungarian notation is neither of these: http://www.joelonsoftware.com/articles/Wrong.html. We should probably call the rule something different. 

Writing the rule is trivial, but I question if you'd actually use this.

**Comment by tommywilkinson** [comment]
Ha! When we were discussing this on our team we actually passed around links to that exact Joel on Software article and had some pretty heated discussions about the place of Apps Hungarian and the use of type inference vs. explicitly typing objects in Typescript and C#. I had a note to link to that article when creating this issue, but completely forgot to include it when the time came to write it out!

To answer numbers one and two (since they seem like they're getting at the same thing), I'd refine what I described above. When we discussed this rule as a team, we wanted two separate functions:

- To disallow type literals as a prefix/suffix on top of a variable name. Based on your examples, we'd definitely have to allow the full type name as being the variable name (those are definitely the most reasonable option in many cases). The potential for this rule being too strict is high, so we also thought about the second form below. 
- Allow users to specify a set of names which are disallow as prefixes or suffixes of variable names. In this case, you could use it to disallow the "s_", "m_", "cu_", etc. This was a pretty crucial caveat we made to the first part, and where the hungarian part of the name really comes from

Our desire for this rule was prompted by the growth of a particular anti-pattern. We added a new WatchableOptionArg type and when people started using them, they used the watchableOptionArg suffix to differentiate between the existing args and the current ones. From there that grew into everybody suffixing any instance of these new WatchableOptionArgs with the type name, even in entirely new classes. Some instances from our code base looked a lot like the following:

```typescript
public interface IButtonArgs{

// Bad. Needlessly suffixing the type
 descriptionWatchableOptionArg: WatchableOptionArg;

// Good! Not repeating information
description: WatchableOptionArg

// Bad. Our team doesn't like prefixes like "m_" and such
m_description: WatchableOptionArg;
}
```

We were in the midst of slowly eradicating pattern this from our code base (we're still working on it some time later) when we thought that it would be useful to have a rule like this that could test for it and help us prevent . We've had similar issues as standards have evolved in our code base but older developers continue to prefix variables with "m_" and "s_" in particular. 

 For number 3, we're a pretty sizeable team of  ~60-70 software developers, so we get all the fun of making this work in a real world code base and having to accommodate different people's styles. The false positive rate would definitely determine whether we used a full scan for the type name, or just use the prefix/suffix checker. If I had to pick, our team would definitely use the self-specified list of disallowed prefixes/suffixes over the checker for literal types in the name.

**Comment by HamletDRC** [comment]
Are these examples correct? 

>> To disallow type literals as a prefix/suffix on top of a variable name. 

    const controller = new Controller();       // OK because variable is exactly the same as type
    const appController = new Controller();    // banned because variable has type name as suffix
    const controllerForApp = new Controller(); // banned because variable has type name as prefix 

    const firstString = someArray[0];  // banned because 'string' is suffix and type
    const firstElement = someArray[0]; // OK because string does not appear in the variable name

    
>> Allow users to specify a set of names which are disallow as prefixes or suffixes of variable names.

assuming the rule is configured with this regex to test rule names:   ^(?!(m|s)_).+
(this is a negative lookahead that will match things to do *not* start with m_ or s_

    m_MyVar; // banned b/c it starts with m_
    s_myVar; // banned b/c is starts with s_
    iMyVar;  // OK
    sMyVar;  // OK
    
Did I capture your statements correctly? 
Are there any other edge cases or examples that I can test?

**Comment by HamletDRC** [comment]
@tommywilkinson any feedback on the above comment? Is this what you need?

**Comment by HamletDRC** [comment]
@tommywilkinson any input here?

**Comment by tommywilkinson** [comment]
Those examples look good to me, although the regex looks like it's missing the "_" after the m and s.  
I would add one more example variable:
```
sum_MyVar; // allowed, since it doesn't start with m_
```
But that might be going too far into just playing regex golf with the particular example regex you chose. 

Also, just to be clear the regex would be the configurable part of the rule, and users could specify any regex there, correct? In our team discussions we always talked about the rules as being based on prefixes and suffixes as a way of avoiding issues where substrings are a part of longer names, but using a regex to configure gives more flexibility than what we originally discussed. Users can configure the regex to be as specific or as loose as they want, which will give the rule some nice flexibility.

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated**; therefore, it and tslint-microsoft-contrib are no longer accepting pull requests for major new changes or features. See #4534. 😱

If you'd like to see this change implemented, you have two choices:
* **Recommended:** Check if this is available in [ESLint](https://eslint.org) + [typescript-eslint](https://typescript-eslint.io) ✅
* _Not Recommended: Fork TSLint or tslint-microsoft-contrib locally_ 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_
